libsolidity/analysis/TypeChecker.cpp

```diff
@@ -673,6 +673,11 @@ bool TypeChecker::visit(FunctionDefinition const& _function)
 				"This type is only supported in the new experimental ABI encoder. "
 				"Use \"pragma experimental ABIEncoderV2;\" to enable the feature."
 			);
+		if (_function.isConstructor() && TypeType(type(*var)) == *m_scope->type())
+			m_errorReporter.typeError(
+				var->location(),
+				"Constructor arguments cannot have the contract itself as type."
```

Comment:
My first hunch was this should be a bug as you did, but after further consideration it may make sense to support this.

There were some examples in the past of splitting/forking contracts, where the reference is to another contract of the same time.